---
title: "L. variegatus Cell Culture - scRNA-seq wit Seurat - 15% FBS"
output: 
  html_document: 
    fig_width: 10
    fig_height: 10
date: '2022-03-10'
name: Kate Castellano
editor_options: 
  chunk_output_type: inline
---
#https://satijalab.org/seurat/articles/merge_vignette.html
#https://satijalab.org/seurat/articles/pbmc3k_tutorial.html

```{r read in file}
library(dplyr)
library(Seurat)
library(patchwork)

Lv.15FBS <- Read10X(data.dir="15FBS_filtered_feature_bc_matrix/")
```

```{r make seurat object}
# Initialize the Seurat object with the raw (non-normalized data).
Lv15FBS <- CreateSeuratObject(counts = Lv.15FBS, project = "15FBS", min.cells = 3, min.features = 200)
Lv15FBS
#An object of class Seurat 
#19728 features across 5478 samples within 1 assay 
#Active assay: RNA (19728 features, 0 variable features)
```

```{r QC,pre-processing}
# The [[ operator can add columns to object metadata. This is a great place to stash QC stats
Lv15FBS[["percent.mt"]] <- PercentageFeatureSet(Lv15FBS, pattern = "^gene-DR311")

# Show QC metrics for the first 5 cells
head(Lv15FBS@meta.data, 5)
```

```{r visualize QC metrics}
# Visualize QC metrics as a violin plot
VlnPlot(Lv15FBS, features = c("nFeature_RNA", "nCount_RNA", "percent.mt"), ncol = 3)

dev.copy(png, "QC_features_15FBScounts.png", width = 900, height = 800)
dev.off()

dev.copy(tiff, "QC_features_15FBScounts.tiff")
dev.off()

dev.copy(svg, "QC_features_15FBScounts.svg")
dev.off()

#QC scatterplot
plot1 <- FeatureScatter(Lv15FBS , feature1 = "nCount_RNA", feature2 = "percent.mt")
plot2 <- FeatureScatter(Lv15FBS , feature1 = "nCount_RNA", feature2 = "nFeature_RNA")
plot1 + plot2
#manually save as TIFF and SVG: QC_features_15FBScounts_scatterplot.png
```

```{r remove outliers}
#based on violin plot, remove super high and super low expressors
#filter out cells that have unique feature counts over 4000 and less than 500 and cells with that have > 30% mitochondrial counts
Lv15FBS <- subset(Lv15FBS, subset = nFeature_RNA > 500 & nFeature_RNA < 4000 & percent.mt < 30)
#went with 30 because it kept a little bit more cells in the analysis (30% cutoff = 3322 total number of cells, no cutoff = 5,398 cells and 20% cutoff = 2919)

#visual violin plots after filtering
VlnPlot(Lv15FBS, features = c("nFeature_RNA", "nCount_RNA", "percent.mt"), ncol = 3)
#manually save as TIFF and SVG: QC_features_15FBScounts_500-400_30mt
```

```{r Normalize}
#LogNormalize: Feature counts for each cell are divided by the total counts for that cell and multiplied by the scale.factor. This is then natural-log transformed using log1p
Lv15FBS <- NormalizeData(Lv15FBS, normalization.method = "LogNormalize", scale.factor = 10000)
```

```{r Identify highly variable features}
#We next calculate a subset of features that exhibit high cell-to-cell variation in the dataset (i.e, they are highly expressed in some cells, and lowly expressed in others). We and others have found that focusing on these genes in downstream analysis helps to highlight biological signal in single-cell datasets.

Lv15FBS <- FindVariableFeatures(Lv15FBS, selection.method = "vst", nfeatures = 2000)

# Identify the 10 most highly variable genes
top10 <- head(VariableFeatures(Lv15FBS), 10)

# plot variable features with and without labels
plot1 <- VariableFeaturePlot(Lv15FBS)
plot2 <- LabelPoints(plot = plot1, points = top10, repel = TRUE)
plot1 + plot2
```

```{r scaling data}
#Next, we apply a linear transformation (‘scaling’) that is a standard pre-processing step prior to dimensional reduction techniques like PCA. The ScaleData() function:

#shifts the expression of each gene, so that the mean expression across cells is 0
#scales the expression of each gene, so that the variance across cells is 1
#This step gives equal weight in downstream analyses, so that highly-expressed genes do not dominate
#The results of this are stored in pbmc[["RNA"]]@scale.data

all.genes <- rownames(Lv15FBS)
Lv15FBS <- ScaleData(Lv15FBS, features = all.genes)
```

```{r linear dimensional reduction}
#Next we perform PCA on the scaled data. By default, only the previously determined variable features are used as input, but can be defined using features argument if you wish to choose a different subset
Lv15FBS <- RunPCA(Lv15FBS, features = VariableFeatures(object = Lv15FBS))

# Examine and visualize PCA results a few different ways
print(Lv15FBS[["pca"]], dims = 1:5, nfeatures = 5)

#visualize PCA
VizDimLoadings(Lv15FBS, dims = 1:2, reduction = "pca")

dev.copy(png, "pca.png", width = 900, height = 800)
dev.off()

dev.copy(pdf, "pca.pdf")
dev.off()

#visualize PCA #2
DimPlot(Lv15FBS, reduction = "pca")

dev.copy(png, "pca_2.png", width = 900, height = 800)
dev.off()

dev.copy(pdf, "pca_2.pdf")
dev.off()

#visualize heatmap
DimHeatmap(Lv15FBS, dims = 1:5, cells = 500, balanced = TRUE)

dev.copy(png, "heatmap.png", width = 900, height = 800)
dev.off()

dev.copy(pdf, "heatmap.pdf")
dev.off()
```

```{r determine dimensionality}
#An alternative heuristic method generates an ‘Elbow plot’: a ranking of principle components based on the percentage of variance explained by each one (ElbowPlot() function). 
#I dont see a clear elbow except maybe right before 15 when it started to level off so chose 15 as my PC cutoff
ElbowPlot(Lv15FBS)

dev.copy(png, "elbowplot_dimensionality.png", width = 900, height = 800)
dev.off()

dev.copy(pdf, "elbowplot_dimensionality.pdf")
dev.off()
```

```{r Cluster the cells}
#picked 16 base don elbow plot
Lv15FBS <- FindNeighbors(Lv15FBS, dims = 1:20)
Lv15FBS <- FindClusters(Lv15FBS, resolution = 0.5)

#Modularity Optimizer version 1.3.0 by Ludo Waltman and Nees Jan van Eck
#Number of nodes: 5398
#Number of edges: 189178

#write data to file
library(data.table)
data_to_write_out <- as.data.frame(as.matrix(Lv15FBS@assays$RNA@scale.data))
fwrite(x = data_to_write_out, row.names = TRUE, file = "LV15FBS_scale.data.csv")
```


```{r non-linear dimensional reduction (tSNE)}
Lv15FBSTSNE <- RunTSNE(Lv15FBS, dims = 1:16)

#visualize t-SNE plot
DimPlot(Lv15FBSTSNE, reduction = "tsne")

dev.copy(png, "cellclusters_15FBS_tSNE.png", width = 900, height = 800)
dev.off()

dev.copy(pdf, "cellClusters_15FBS_tSNE.pdf")
dev.off()

DimPlot(Lv15FBSTSNE, reduction = "tsne", label = TRUE)
dev.copy(png, "cellclusters_15FBS_tSNE_label.png", width = 900, height = 800)
dev.off()

dev.copy(pdf, "cellClusters_15FBS_tSNE_label.pdf")
dev.off()
```
```{r }
###Final Cell Marker Graphs###

#Pigment: homeobox_protein_six1-like, probable_polyketide_synthase_1
VlnPlot(Lv15FBS, features = c("LOC121418887", "LOC121428528"))

dev.copy(png, "violinPlot_15FBS_pigmentMarkers.png", width = 900, height = 500)
dev.off()

dev.copy(svg, "violinPlot_15FBS_pigmentMarkers.svg", width = 10, height = 5)
dev.off()

dev.copy(tiff, "violinPlot_15FBS_pigmentmMarkers.tiff", width = 900, height = 500)
dev.off()

FeaturePlot(Lv15FBSTSNE, features = c("LOC121418887", "LOC121428528"))

dev.copy(png, "featurePlot_15FBS_pigmentMarkers.png", width = 900, height = 500)
dev.off()

dev.copy(svg, "featurePlot_15FBS_pigmentMarkers.svg", width = 10, height = 5)
dev.off()

dev.copy(tiff, "featurePlot_15FBS_pigmentmMarkers.tiff", width = 900, height = 500)
dev.off()

##Stemness Markers##
#protein_maelstrom_homolog, piwi-like_protein_1, piwi-like_protein_1, eEF1A_lysine_and_N-terminal_methyltransferase-like, protein_nanos-like, probable_ATP-dependent_RNA_helicase_DDX4, probable_ATP-dependent_RNA_helicase_DDX4

VlnPlot(Lv15FBS, features = c("LOC121420310", "LOC121408762", "LOC121431753", "LOC121428667", "LOC121408828", "LOC121423730", "LOC121424246"))

dev.copy(png, "violinPlot_15FBS_stemMarkers.png", width = 900, height = 700)
dev.off()

dev.copy(svg, "violinPlot_15FBS_stemMarkers.svg", width = 10, height = 10)
dev.off()

dev.copy(tiff, "violinPlot_15FBS_stemMarkers.tiff", width = 900, height = 700)
dev.off()

FeaturePlot(Lv15FBSTSNE, features = c("LOC121420310", "LOC121408762", "LOC121431753", "LOC121428667", "LOC121408828", "LOC121423730", "LOC121424246"))
dev.copy(png, "featurePlot_15FBS_stemMarkers.png", width = 900, height = 800)
dev.off()

dev.copy(svg, "featurePlot_15FBS_stemMarkers.svg", width = 10, height = 10)
dev.off()

dev.copy(tiff, "featurePlot_15FBS_stemMarkers.tiff", width = 900, height = 800)
dev.off()

##Proliferation Markers##
#proliferating_cell_nuclear_antigen-like, myc_protein-like, telomerase_reverse_transcriptase-like, telomerase_protein_component_1-like, telomerase_protein_component_1-like, PIN2/TERF1-interacting_telomerase_inhibitor_1-like, PIN2/TERF1-interacting_telomerase_inhibitor_1-like, G2/mitotic-specific_cyclin-A-like, G2/mitotic-specific_cyclin-A-like
VlnPlot(Lv15FBS, features = c("LOC121431754", "LOC121430677", "LOC121429790", "LOC121412153", "LOC121424081", "LOC121417261", "LOC121417323", "LOC121410426", "LOC121409442"))

dev.copy(png, "violinPlot_15FBS_prolifMarkers_1.png", width = 900, height = 700)
dev.off()

dev.copy(svg, "violinPlot_15FBS_prolifMarkers_1.svg", width = 10, height = 10)
dev.off()

dev.copy(tiff, "violinPlot_15FBS_prolifmMarkers_1.tiff", width = 900, height = 700)
dev.off()


FeaturePlot(Lv15FBSTSNE, features = c("LOC121431754", "LOC121430677", "LOC121429790", "LOC121412153", "LOC121424081", "LOC121417261", "LOC121417323", "LOC121410426", "LOC121409442"))

dev.copy(png, "featurePlot_15FBS_prolifMarkers_1.png", width = 900, height = 800)
dev.off()

dev.copy(svg, "featurePlot_15FBS_prolifMarkers_1.svg", width = 10, height = 10)
dev.off()

dev.copy(tiff, "featurePlot_15FBS_prolifmMarkers_1.tiff", width = 900, height = 800)
dev.off()


#G2/mitotic-specific_cyclin-B-like, G2/mitotic-specific_cyclin-B3-like, G2/mitotic-specific_cyclin_S13-7-like, G1/S-specific_cyclin-D1-like, G1/S-specific_cyclin-E, G1/S-specific_cyclin-D1-like, cyclin-D-binding_Myb-like_transcription_factor_1, cyclin-D-binding_Myb-like_transcription_factor_1, G1/S-specific_cyclin-E, S_phase_cyclin_A-associated_protein_in_the_endoplasmic_reticulum-like
VlnPlot(Lv15FBS, features = c("LOC121416277", "LOC121429467", "LOC121419378", "LOC121405827", "LOC121408084", "LOC121405827", "LOC121410996", "LOC121413797", "LOC121408084", "LOC121415132"))

dev.copy(png, "violinPlot_15FBS_prolifMarkers_2.png", width = 900, height = 800)
dev.off()

dev.copy(svg, "violinPlot_15FBS_prolifMarkers_2.svg", width = 10, height = 10)
dev.off()

dev.copy(tiff, "violinPlot_15FBS_prolifmMarkers_2.tiff", width = 900, height = 800)
dev.off()

FeaturePlot(Lv15FBSTSNE, features = c("LOC121416277", "LOC121429467", "LOC121419378", "LOC121405827", "LOC121408084", "LOC121405827", "LOC121410996", "LOC121413797", "LOC121408084", "LOC121415132"))

dev.copy(png, "featurePlot_15FBS_prolifMarkers_2.png", width = 900, height = 800)
dev.off()

dev.copy(svg, "featurePlot_15FBS_prolifMarkers_2.svg", width = 10, height = 10)
dev.off()

dev.copy(tiff, "featurePlot_15FBS_prolifmMarkers_2.tiff", width = 900, height = 800)
dev.off()


##Neuronal Markers##
#notch, synaptotagmin-1-like, synaptotagmin-1-like, synaptotagmin-6-like, synaptotagmin-12-like, synaptotagmin-14-like, neurofilament_heavy_polypeptide-like, neurofilament_heavy_polypeptide-like, neurofilament_heavy_polypeptide-like, neurofilament_heavy_polypeptide-like

VlnPlot(Lv15FBS, features = c("LOC447795", "LOC121410540", "LOC121405894", "LOC121405860", "LOC121406212", "LOC121418961", "LOC121414907", "LOC121424392", "LOC121424392", "LOC121424392"))

dev.copy(png, "violinPlot_15FBS_neuronalMarkers_1.png", width = 900, height = 800)
dev.off()

dev.copy(svg, "violinPlot_15FBS_neuronalMarkers_1.svg", width = 10, height = 10)
dev.off()

dev.copy(tiff, "violinPlot_15FBS_neuronalMarkers_1.tiff", width = 900, height = 800)
dev.off()


FeaturePlot(Lv15FBSTSNE, features = c("LOC447795", "LOC121410540", "LOC121405894", "LOC121405860", "LOC121406212", "LOC121418961", "LOC121414907", "LOC121424392", "LOC121424392", "LOC121424392"))

dev.copy(png, "featurePlot_15FBS_neuronalMarkers_1.png", width = 900, height = 800)
dev.off()

dev.copy(svg, "featurePlot_15FBS_neuronalMarkers_1.svg", width = 10, height = 10)
dev.off()

dev.copy(tiff, "featurePlot_15FBS_neuronalMarkers_1.tiff", width = 900, height = 800)
dev.off()

#neurofilament_heavy_polypeptide-like, neurofilament_heavy_polypeptide-like, neurofilament_medium_polypeptide-like, metabotropic_glutamate_receptor_3-like, hyalin, hyalin-like, hyalin-like, hyalin-like, neuronal_acetylcholine_receptor_subunit_alpha-3-like, neural_proliferation_differentiation_and_control_protein_1-like

VlnPlot(Lv15FBS, features = c("LOC121424392", "LOC121432068", "LOC121425197", "LOC121429617", "LOC446193", "LOC121419137", "LOC121421150", "LOC121425550", "LOC121429806", "LOC121429231"))

dev.copy(png, "violinPlot_15FBS_neuronalMarkers_2.png", width = 900, height = 800)
dev.off()

dev.copy(svg, "violinPlot_15FBS_neuronalMarkers_2.svg", width = 10, height = 10)
dev.off()

dev.copy(tiff, "violinPlot_15FBS_neuronalmarkers_2.tiff", width = 900, height = 800)
dev.off()


FeaturePlot(Lv15FBSTSNE, features = c("LOC121424392", "LOC121432068", "LOC121425197", "LOC121429617", "LOC446193", "LOC121419137", "LOC121421150", "LOC121425550", "LOC121429806", "LOC121429231"))

dev.copy(png, "featurePlot_15FBS_neuronalMarkers_2.png", width = 900, height = 800)
dev.off()

dev.copy(svg, "featurePlot_15FBS_neuronalMarkers_2.svg", width = 10, height = 10)
dev.off()

dev.copy(tiff, "featurePlot_15FBS_neuronalmarkers_2.tiff", width = 900, height = 800)
dev.off()

##ectoderm markers##
#transcription_factor_HES-1-like, transcription_factor_HES-1-A-like, transcription_factor_HES-4-like, chordin-like

VlnPlot(Lv15FBS, features = c("LOC121427522", "LOC121427235", "LOC121427575", "LOC121427607"))
dev.copy(png, "violinPlot_15FBS_ectodermlMarkers.png", width = 900, height = 800)
dev.off()

dev.copy(svg, "violinPlot_15FBS_ectodermMarkers.svg", width = 10, height = 10)
dev.off()

dev.copy(tiff, "violinPlot_15FBS_ectodermMarkers.tiff", width = 900, height = 800)
dev.off()

FeaturePlot(Lv15FBSTSNE, features = c("LOC121427522", "LOC121427235", "LOC121427575", "LOC121427607"))
dev.copy(png, "featurePlot_15FBS_ectodermlMarkers.png", width = 900, height = 800)
dev.off()

dev.copy(svg, "featurePlot_15FBS_ectodermMarkers.svg", width = 10, height = 10)
dev.off()

dev.copy(tiff, "featurePlot_15FBS_ectodermMarkers.tiff", width = 900, height = 800)
dev.off()

##Skeleton Markers##
#mesenchyme-specific_cell_surface_glycoprotein-like, mesenchyme-specific_cell_surface_glycoprotein-like, mesenchyme-specific_cell_surface_glycoprotein-like, 34_kDa_spicule_matrix_protein, 41_kDa_spicule_matrix_protein-like, 30_kDa_spicule_matrix_protein_alpha-like
VlnPlot(Lv15FBS, features = c("LOC121428121", "LOC121428213", "LOC121427215", "LOC121405815", "LOC121405817", "LOC121406020"))

dev.copy(png, "violinPlot_15FBS_skeletonMarkers_1.png", width = 900, height = 800)
dev.off()

dev.copy(svg, "violinPlot_15FBS_skeletonMarkers_1.svg", width = 10, height = 10)
dev.off()

dev.copy(tiff, "violinPlot_15FBS_skeletonMarkers_1.tiff", width = 900, height = 800)
dev.off()


FeaturePlot(Lv15FBSTSNE, features = c("LOC121428121", "LOC121428213", "LOC121427215", "LOC121405815", "LOC121405817", "LOC121406020"))

dev.copy(png, "featurePlot_15FBS_skeletonMarkers_1.png", width = 900, height = 800)
dev.off()

dev.copy(svg, "featurePlot_15FBS_skeletonMarkers_1.svg", width = 10, height = 10)
dev.off()

dev.copy(tiff, "featurePlot_15FBS_skeletonMarkers_1.tiff", width = 900, height = 800)
dev.off()

#34_kDa_spicule_matrix_protein-like, 27_kDa_primary_mesenchyme-specific_spicule_protein-like, carbonic_anhydrase_1-like, carbonic_anhydrase-related_protein_10-like, beta_carbonic_anhydrase_1-like, carbonic_anhydrase_14-like
VlnPlot(Lv15FBS, features = c("LOC121406031", "LOC121406034", "LOC121406846", "LOC121429170", "LOC121429566", "LOC121430996"))

dev.copy(png, "violinPlot_15FBS_skeletonMarkers_2.png", width = 900, height = 800)
dev.off()

dev.copy(svg, "violinPlot_15FBS_skeletonMarkers_2.svg", width = 10, height = 10)
dev.off()

dev.copy(tiff, "violinPlot_15FBS_skeletonMarkers_2.tiff", width = 900, height = 800)
dev.off()


FeaturePlot(Lv15FBSTSNE, features = c("LOC121406031", "LOC121406034", "LOC121406846", "LOC121429170", "LOC121429566", "LOC121430996"))

dev.copy(png, "featurePlot_15FBS_skeletonMarkers_2.png", width = 900, height = 800)
dev.off()

dev.copy(svg, "featurePlot_15FBS_skeletonMarkers_2.svg", width = 10, height = 10)
dev.off()

dev.copy(tiff, "featurePlot_15FBS_skeletonMarkers_2.tiff", width = 900, height = 800)
dev.off()

##Muscle Markers##
#myosin_heavy_chain,_striated_muscle-like, myosin_light_chain_kinase,_smooth_muscle-like, myosin_light_chain_kinase,_smooth_muscle-like, broad_substrate_specificity_ATP-binding_cassette_transporter_ABCG2-like, myoblast_determination_protein_1_homolog

VlnPlot(Lv15FBS, features = c("LOC121408547", "LOC121424934", "LOC121425215", "LOC121431580", "LOC121405674"))
dev.copy(png, "violinPlot_15FBS_muscleMarkers.png", width = 900, height = 800)
dev.off()

dev.copy(svg, "violinPlot_15FBS_musclenMarkers.svg", width = 10, height = 10)
dev.off()

dev.copy(tiff, "violinPlot_15FBS_muscleMarkers.tiff", width = 900, height = 800)
dev.off()


FeaturePlot(Lv15FBSTSNE, features = c("LOC121408547", "LOC121424934", "LOC121425215", "LOC121431580", "LOC121405674"))
dev.copy(png, "featurePlot_15FBS_muscleMarkers.png", width = 900, height = 800)
dev.off()

dev.copy(svg, "featurePlot_15FBS_musclenMarkers.svg", width = 10, height = 10)
dev.off()

dev.copy(tiff, "featurePlot_15FBS_muscleMarkers.tiff", width = 900, height = 800)
dev.off()


##Cilia Markers ##
#tektin-1-like, tektin-4-like, tektin-3-like, tektin-B1-like

VlnPlot(Lv15FBS, features = c("LOC121408954", "LOC121423108", "LOC121423323", "LOC121430708"))

dev.copy(png, "violinPlot_15FBS_ciliaMarkers.png", width = 900, height = 800)
dev.off()

dev.copy(svg, "violinPlot_15FBS_ciliaMarkers.svg", width = 10, height = 10)
dev.off()

dev.copy(tiff, "violinPlot_15FBS_ciliaMarkers.tiff", width = 900, height = 800)
dev.off()


FeaturePlot(Lv15FBSTSNE, features = c("LOC121408954", "LOC121423108", "LOC121423323", "LOC121430708"))

dev.copy(png, "featurePlot_15FBS_ciliaMarkers.png", width = 900, height = 800)
dev.off()

dev.copy(svg, "featurePlot_15FBS_ciliaMarkers.svg", width = 10, height = 10)
dev.off()

dev.copy(tiff, "featurePlot_15FBS_ciliaMarkers.tiff", width = 900, height = 800)
dev.off()


##Receptors##

#EGF
VlnPlot(Lv15FBS, features = c("LOC121405695", "LOC121421295", "LOC121429013"))

dev.copy(png, "violinPlot_15FBS_EGF.png", width = 900, height = 800)
dev.off()

dev.copy(svg, "violinPlot_15FBS_EGF.svg", width = 10, height = 10)
dev.off()

dev.copy(tiff, "violinPlot_15FBS_EGF.tiff", width = 900, height = 800)
dev.off()


FeaturePlot(Lv15FBSTSNE, features = c("LOC121405695", "LOC121421295", "LOC121429013"))

dev.copy(png, "featurePlot_15FBS_EGF.png", width = 900, height = 800)
dev.off()

dev.copy(svg, "featurePlot_15FBS_EGF.svg", width = 10, height = 10)
dev.off()

dev.copy(tiff, "featurePlot_15FBS_EGF.tiff", width = 900, height = 800)
dev.off()

#FGF
VlnPlot(Lv15FBS, features = c("LOC121415946", "LOC121419550", "LOC121425416", "LOC121425598", "LOC121425661", "LOC121429921", "LOC121431286"))

dev.copy(png, "violinPlot_15FBS_FGF.png", width = 900, height = 800)
dev.off()

dev.copy(svg, "violinPlot_15FBS_FGF.svg", width = 10, height = 10)
dev.off()

dev.copy(tiff, "violinPlot_15FBS_FGF.tiff", width = 900, height = 800)
dev.off()


FeaturePlot(Lv15FBSTSNE, features = c("LOC121415946", "LOC121419550", "LOC121425416", "LOC121425598", "LOC121425661", "LOC121429921", "LOC121431286"))

dev.copy(png, "featurePlot_15FBS_FGF.png", width = 900, height = 800)
dev.off()

dev.copy(svg, "featurePlot_15FBS_FGF.svg", width = 10, height = 10)
dev.off()

dev.copy(tiff, "featurePlot_15FBS_FGF.tiff", width = 900, height = 800)
dev.off()

#hepatocyte
#all hepatocyte annotations
#VlnPlot(Lv15FBS, features = c("LOC121407560", "LOC121407763", "LOC121408778", "LOC121410774", "LOC121419318", "LOC121422516", "LOC121426646", "LOC121428657"))

#hepatocyte receptors
VlnPlot(Lv15FBS, features = c("LOC121410774", "LOC121422516", "LOC121426646"))
dev.copy(png, "violinPlot_15FBS_hepatocyte.png", width = 900, height = 700)
dev.off()

dev.copy(svg, "violinPlot_15FBS_hepatocyte.svg", width = 10, height = 10)
dev.off()

dev.copy(tiff, "violinPlot_15FBS_hepatocyte.tiff", width = 900, height = 800)
dev.off()


FeaturePlot(Lv15FBSTSNE, features = c("LOC121410774", "LOC121422516", "LOC121426646"))
dev.copy(png, "featurePlot_15FBS_hepatocyte.png", width = 900, height = 700)
dev.off()

dev.copy(svg, "featurePlot_15FBS_hepatocyte.svg", width = 10, height = 10)
dev.off()

dev.copy(pdf, "featurePlot_15FBS_hepatocyte.pdf", width = 900, height = 800)
dev.off()
```

```{r}
#Get the number of cells in each cluster
library(data.table)
library(magrittr)
library(dplyr)

#get the meta data and put it in a table
md <- Lv15FBSTSNE@meta.data %>% as.data.table
head(md)

#get the total number of cells in each cluster
table(Lv15FBSTSNE$seurat_clusters)
```

```{r  Function to calculate percentage of cells expressing marker genes}
# updated 1/31/2020 to accommodate V3.1
# updated 2/4/2020 to output "NA" for genes not detected in certain subgroups
#Function to get the percent of cells expressing a certain gene in each cluster
#https://github.com/satijalab/seurat/issues/371

PrctCellExpringGene <- function(object, genes, group.by = "all"){
    if(group.by == "all"){
        prct = unlist(lapply(genes,calc_helper, object=object))
        result = data.frame(Markers = genes, Cell_proportion = prct)
        return(result)
    }

    else{        
        list = SplitObject(object, group.by)
        factors = names(list)

        results = lapply(list, PrctCellExpringGene, genes=genes)
        for(i in 1:length(factors)){
        results[[i]]$Feature = factors[i]
        }
        combined = do.call("rbind", results)
        return(combined)
    }
}

calc_helper <- function(object,genes){
    counts = object[['RNA']]@counts
    ncells = ncol(counts)
    if(genes %in% row.names(counts)){
    sum(counts[genes,]>0)/ncells*100
    }else{return(NA)}
}

```

```{r  calculate percentage of cells expressing marker genes}
#calculate the percentage of cells in ech cluster expressing certain genes (need function above)

#To get the number of cells you can multiply the percentage by the total number of cells in each cluster (calculate above). You can also subset cells expressing a gene into another table and calculate the number of cells in each cluster using this code (I checked that all the numbers work out):
  # cycllinD_subset <- subset(x = Lv15FBSTSNE, subset = LOC121405827 > 0)
  # table(cycllinD_subset$seurat_clusters)

#proliferation markers
d <- PrctCellExpringGene(Lv15FBSTSNE ,genes =c("LOC121409442","LOC121416277","LOC121405827","LOC121408084","LOC121415132", "LOC121410761", "LOC447791", "LOC121409919", "LOC121426230", "LOC121431754", "LOC121430677", "LOC121429790"), group.by = "seurat_clusters")
write.csv(d, "15FBSPercCells_ProlifMarkers_2023Oct.csv", row.names=FALSE)

#ectoderm
d <- PrctCellExpringGene(Lv15FBSTSNE ,genes =c("LOC121427522", "LOC121427235", "LOC121427575", "LOC121427607", "LOC121420667", "LOC446163", "LOC121410373", "LOC446159", "LOC121430393", "LOC447795", "LOC121405894", "LOC121418961", "LOC121424392", "LOC121425197", "LOC121429617", "LOC446193", "LOC121429806", "LOC121429231"), group.by = "seurat_clusters")
write.csv(d, "15FBSPercCells_ectodermMarkers_2023Oct.csv", row.names=FALSE)

#endoderm
d <- PrctCellExpringGene(Lv15FBSTSNE ,genes =c("LOC121406173", "LOC121428657", "LOC121419318"), group.by = "seurat_clusters")
write.csv(d, "15FBSPercCells_endodermMarkers_2023Oct.csv", row.names=FALSE)

#mesoderm
d <- PrctCellExpringGene(Lv15FBSTSNE ,genes =c("LOC121408547", "LOC121424934", "LOC121425215", "LOC121406031", "LOC121406034", "LOC121406020", "LOC121406031", "LOC121428121"), group.by = "seurat_clusters")
write.csv(d, "15FBSPercCells_mesodermMarkers_2023Oct.csv", row.names=FALSE)

#germ cell
d <- PrctCellExpringGene(Lv15FBSTSNE ,genes =c("LOC121408828", "LOC121408762", "LOC121431753", "LOC121424246", "LOC121420310"), group.by = "seurat_clusters")
write.csv(d, "15FBSPercCells_germcellMarkers_2023Oct.csv", row.names=FALSE)

#pigment
d <- PrctCellExpringGene(Lv15FBSTSNE ,genes =c("LOC121428528" ,"LOC121418887"), group.by = "seurat_clusters")
write.csv(d, "15FBSPercCells_pigmentMarkers_2023Oct.csv", row.names=FALSE)

#cillia
d <- PrctCellExpringGene(Lv15FBSTSNE ,genes =c("LOC121408954" ,"LOC121423108" ,"LOC121423323" ,"LOC121430708"), group.by = "seurat_clusters")
write.csv(d, "15FBSPercCells_cilliaMarkers_2023Oct.csv", row.names=FALSE)

#DNA polymerase genes
d <- PrctCellExpringGene(Lv15FBSTSNE ,genes =c("LOC121429881", "LOC121408448", "LOC121428701"), group.by = "seurat_clusters")
write.csv(d, "15FBSPercCells_dnapolyMarkers_2023Oct.csv", row.names=FALSE)
```


```{r }
# This loop just runs the FindMarkers function on all of the clusters
lapply(
  levels(Lv15FBSTSNE[["seurat_clusters"]][[1]]),
  function(x)FindMarkers(Lv15FBSTSNE,ident.1 = x,min.pct = 0.25)
) -> cluster.markers

# This adds the cluster number to the results of FindMarkers
sapply(0:(length(cluster.markers)-1),function(x) {
  cluster.markers[[x+1]]$gene <<- rownames(cluster.markers[[x+1]])
  cluster.markers[[x+1]]$cluster <<- x
})

# Finally we collapse the list of hits down to a single table and sort it by FDR to put the most significant ones first
as_tibble(do.call(rbind,cluster.markers)) %>% arrange(p_val_adj) -> cluster.markers
cluster.markers

#prints to file
write.csv(cluster.markers, "15FBS_allClusters_Genes_2024March.csv")
```

```{r }
#Get counts for each gene in each cluster
setwd("C:/Users/kate.castellano/OneDrive - Gloucester Marine Genomics Institute/Documents/KCastellano/SeaUrchin/SeaUrchin_Data/10x_seaurchin/Seurat_15FBS/2024_MarchAnalysis_mitoFilter")
matrix <- AverageExpression(object = Lv15FBSTSNE)$RNA
write.csv(matrix, "Lv_15FBS_GeneCount_allClustersNEW.csv")
getwd()
```

```{r }
#get average expression across all clusters for cell cycle genes (to compare to other conditions)

markers <- AverageExpression(object = Lv15FBSTSNE, group.by = c('orig.ident'))$RNA

#get specific cell cycle genes
my_markers <- markers[c("LOC121409442", "LOC121416277", "LOC121405827", "LOC121408084", "LOC121415132", "LOC121429790", "LOC121430677", "LOC121431754", "LOC121429881", "LOC121408448", "LOC121428701", "LOC121408828", "LOC121408762", "LOC121431753", "LOC121424246", "LOC121420310"), ]
my_markers
```